File: cua-helper/app/src/main/java/com/hermes/cua/CuaAccessibilityService.java
package com.hermes.cua;

import android.accessibilityservice.AccessibilityService;
import android.accessibilityservice.GestureDescription;
import android.graphics.Bitmap;
import android.graphics.ColorSpace;
import android.graphics.Path;
import android.hardware.HardwareBuffer;
import android.os.Bundle;
import android.os.Build;
import android.view.accessibility.AccessibilityEvent;
import android.view.accessibility.AccessibilityNodeInfo;

import java.io.ByteArrayOutputStream;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class CuaAccessibilityService extends AccessibilityService {

    @Override
    public void onServiceConnected() {
        super.onServiceConnected();
        CuaService.setAccService(this);
        android.util.Log.i("CuaAcc", "Service connected OK");
    }

    @Override
    public void onAccessibilityEvent(AccessibilityEvent event) {}

    @Override
    public void onInterrupt() {}

    public void tap(int x, int y) {
        try {
            Path path = new Path();
            path.moveTo(x, y);
            GestureDescription.Builder builder = new GestureDescription.Builder();
            builder.addStroke(new GestureDescription.StrokeDescription(path, 0, 50));
            dispatchGesture(builder.build(), null, null);
        } catch (Exception e) {
            android.util.Log.e("CuaAcc", "tap failed", e);
        }
    }

    public void swipe(int x1, int y1, int x2, int y2, int durMs) {
        try {
            Path path = new Path();
            path.moveTo(x1, y1);
            path.lineTo(x2, y2);
            GestureDescription.Builder builder = new GestureDescription.Builder();
            builder.addStroke(new GestureDescription.StrokeDescription(path, 0, durMs));
            dispatchGesture(builder.build(), null, null);
        } catch (Exception e) {
            android.util.Log.e("CuaAcc", "swipe failed", e);
        }
    }

    public String dumpUi() {
        try {
            AccessibilityNodeInfo root = getRootInActiveWindow();
            if (root == null) return "<empty/>";
            String result = dumpNodeSimple(root);
            root.recycle();
            return result;
        } catch (Exception e) {
            return "{\"error\":\"" + e.getMessage() + "\"}";
        }
    }

    private String dumpNodeSimple(AccessibilityNodeInfo node) {
        // Minimal XML dump without deep recursion overhead
        StringBuilder sb = new StringBuilder();
        dumpNodeXml(node, sb, 0, 50); // max depth 50
        return sb.toString();
    }

    private void dumpNodeXml(AccessibilityNodeInfo node, StringBuilder sb, int depth, int maxDepth) {
        if (node == null || depth >= maxDepth) return;
        for (int i = 0; i < depth; i++) sb.append("  ");
        sb.append("<node");
        appendAttr(sb, "class", node.getClassName());
        appendAttr(sb, "text", node.getText());
        appendAttr(sb, "desc", node.getContentDescription());
        appendAttr(sb, "id", node.getViewIdResourceName());
        android.graphics.Rect bounds = new android.graphics.Rect();
        node.getBoundsInScreen(bounds);
        sb.append(" bounds=\"[").append(bounds.left).append(",").append(bounds.top)
          .append("][").append(bounds.right).append(",").append(bounds.bottom).append("]\"");
        sb.append(" clickable=\"").append(node.isClickable()).append("\"");
        
        int childCount = node.getChildCount();
        if (childCount == 0) {
            sb.append("/>\n");
        } else {
            sb.append(">\n");
            for (int i = 0; i < childCount; i++) {
                AccessibilityNodeInfo child = node.getChild(i);
                dumpNodeXml(child, sb, depth + 1, maxDepth);
                if (child != null) child.recycle();
            }
            for (int i = 0; i < depth; i++) sb.append("  ");
            sb.append("</node>\n");
        }
    }

    private void appendAttr(StringBuilder sb, String name, CharSequence value) {
        if (value != null && value.length() > 0) {
            String s = value.toString().replace("&", "&amp;").replace("\"", "&quot;")
                .replace("<", "&lt;").replace(">", "&gt;").replace("\n", "\\n");
            sb.append(" ").append(name).append("=\"").append(s).append("\"");
        }
    }

    public void goBack() {
        performGlobalAction(GLOBAL_ACTION_BACK);
    }

    public void goHome() {
        performGlobalAction(GLOBAL_ACTION_HOME);
    }

    public void inputText(String text) {
        try {
            AccessibilityNodeInfo focused = findFocus(AccessibilityNodeInfo.FOCUS_INPUT);
            if (focused == null) {
                AccessibilityNodeInfo root = getRootInActiveWindow();
                if (root != null) {
                    focused = root.findFocus(AccessibilityNodeInfo.FOCUS_INPUT);
                    root.recycle();
                }
            }
            if (focused != null) {
                if (focused.isEditable()) {
                    Bundle args = new Bundle();
                    args.putCharSequence(AccessibilityNodeInfo.ACTION_ARGUMENT_SET_TEXT_CHARSEQUENCE, text);
                    focused.performAction(AccessibilityNodeInfo.ACTION_SET_TEXT, args);
                }
                focused.recycle();
            }
        } catch (Exception e) {
            android.util.Log.e("CuaAcc", "inputText failed", e);
        }
    }

    public byte[] captureScreen() {
        if (Build.VERSION.SDK_INT < 34) return null;
        try {
            final Bitmap[] result = new Bitmap[1];
            final CountDownLatch latch = new CountDownLatch(1);
            takeScreenshot(
                1, // TAKE_SCREENSHOT_FULL_SCREEN (@hide, value=1)
                Executors.newSingleThreadExecutor(),
                new TakeScreenshotCallback() {
                    @Override
                    public void onSuccess(ScreenshotResult sr) {
                        try {
                            HardwareBuffer hb = sr.getHardwareBuffer();
                            if (hb != null) {
                                Bitmap bmp = Bitmap.wrapHardwareBuffer(hb,
                                    ColorSpace.get(ColorSpace.Named.SRGB));
                                if (bmp != null) {
                                    result[0] = bmp.copy(bmp.getConfig(), bmp.isMutable());
                                    bmp.recycle();
                                }
                                hb.close();
                            }
                        } catch (Exception e) {
                            android.util.Log.e("CuaAcc", "captureScreen fail", e);
                        }
                        latch.countDown();
                    }
                    @Override
                    public void onFailure(int errorCode) {
                        android.util.Log.e("CuaAcc", "captureScreen error " + errorCode);
                        latch.countDown();
                    }
                }
            );
            latch.await(5, TimeUnit.SECONDS);
            if (result[0] != null) {
                ByteArrayOutputStream baos = new ByteArrayOutputStream();
                result[0].compress(Bitmap.CompressFormat.PNG, 90, baos);
                result[0].recycle();
                return baos.toByteArray();
            }
        } catch (Exception e) {
            android.util.Log.e("CuaAcc", "captureScreen exception", e);
        }
        return null;
    }

    public boolean pasteClipboard() {
        try {
            AccessibilityNodeInfo focused = findFocus(AccessibilityNodeInfo.FOCUS_INPUT);
            if (focused == null) {
                AccessibilityNodeInfo root = getRootInActiveWindow();
                if (root != null) {
                    focused = root.findFocus(AccessibilityNodeInfo.FOCUS_INPUT);
                    root.recycle();
                }
            }
            if (focused != null) {
                boolean ok = focused.performAction(AccessibilityNodeInfo.ACTION_PASTE);
                focused.recycle();
                return ok;
            }
        } catch (Exception e) {
            android.util.Log.e("CuaAcc", "paste failed", e);
        }
        return false;
    }
}

File: cua-helper/app/src/main/java/com/hermes/cua/CuaAccessibilityService.kt
package com.hermes.cua

import android.accessibilityservice.AccessibilityService
import android.accessibilityservice.GestureDescription
import android.content.Intent
import android.graphics.Path
import android.graphics.Rect
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.view.accessibility.AccessibilityEvent
import android.view.accessibility.AccessibilityNodeInfo
import java.io.StringWriter
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult

class CuaAccessibilityService : AccessibilityService() {

    companion object {
        var instance: CuaAccessibilityService? = null
            private set
    }

    private val handler = Handler(Looper.getMainLooper())

    override fun onServiceConnected() {
        super.onServiceConnected()
        instance = this
    }

    override fun onAccessibilityEvent(event: AccessibilityEvent?) {
        // We don't need to react to events, just provide services
    }

    override fun onInterrupt() {
        // Service interrupted
    }

    override fun onDestroy() {
        super.onDestroy()
        instance = null
    }

    /**
     * Take a screenshot using the AccessibilityService API (Android 14+).
     */
    fun takeScreenshot() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) { // Android 14
            takeScreenshot(
                SCREENSHOT_CHROME_HIDE_SCREENSHOT,
                object : TakeScreenshotCallback {
                    override fun onScreenshotTaken(screenshot: ScreenshotResult) {
                        val hardwareBuffer = screenshot.hardwareBuffer
                        if (hardwareBuffer != null) {
                            val bitmap = android.graphics.Bitmap.wrapHardwareBuffer(
                                hardwareBuffer, null
                            )
                            val stream = java.io.ByteArrayOutputStream()
                            bitmap?.compress(
                                android.graphics.Bitmap.CompressFormat.PNG, 90, stream
                            )
                            CuaService.lastScreenshotBytes = stream.toByteArray()
                            hardwareBuffer.close()
                        }
                    }

                    override fun onScreenshotFailed() {
                        // Failed to take screenshot
                    }
                },
                handler
            )
        }
    }

    /**
     * Perform a tap gesture at (x, y).
     */
    fun performTap(x: Float, y: Float) {
        val path = Path().apply {
            moveTo(x, y)
        }
        val gesture = GestureDescription.Builder()
            .addStroke(GestureDescription.StrokeDescription(path, 0, 100))
            .build()
        dispatchGesture(gesture, null, null)
    }

    /**
     * Perform a swipe gesture.
     */
    fun performSwipe(x1: Float, y1: Float, x2: Float, y2: Float) {
        val path = Path().apply {
            moveTo(x1, y1)
            lineTo(x2, y2)
        }
        val gesture = GestureDescription.Builder()
            .addStroke(GestureDescription.StrokeDescription(path, 0, 300))
            .build()
        dispatchGesture(gesture, null, null)
    }

    /**
     * Type text using AccessibilityService's ability to set text.
     * This works when a text field is focused.
     */
    fun performText(text: String) {
        val root = rootInActiveWindow ?: return
        val focused = root.findFocus(AccessibilityNodeInfo.FOCUS_INPUT)
        if (focused != null) {
            val args = android.os.Bundle()
            // Use ACTION_SET_TEXT for full replace
            args.putCharSequence(
                AccessibilityNodeInfo.ACTION_ARGUMENT_SET_TEXT_CHARSEQUENCE,
                text
            )
            focused.performAction(
                AccessibilityNodeInfo.ACTION_SET_TEXT,
                args
            )
        }
        root.recycle()
    }

    /**
     * Perform a global action (back, home, recents).
     */
    fun performGlobalAction(action: Int) {
        performGlobalAction(
            when (action) {
                android.accessibilityservice.GestureDescription.GESTURE_GLOBAL_ACTION_BACK
                    -> GLOBAL_ACTION_BACK
                android.accessibilityservice.GestureDescription.GESTURE_GLOBAL_ACTION_HOME
                    -> GLOBAL_ACTION_HOME
                android.accessibilityservice.GestureDescription.GESTURE_GLOBAL_ACTION_RECENTS
                    -> GLOBAL_ACTION_RECENTS
                android.accessibilityservice.GestureDescription.GESTURE_GLOBAL_ACTION_NOTIFICATIONS
                    -> GLOBAL_ACTION_NOTIFICATIONS
                else -> GLOBAL_ACTION_BACK
            }
        )
    }

    /**
     * Get the current UI hierarchy as XML.
     */
    fun getUiHierarchy(): String? {
        val root = rootInActiveWindow ?: return null
        return try {
            val doc = DocumentBuilderFactory.newInstance().newDocumentBuilder().newDocument()
            val rootElement = doc.createElement("hierarchy")
            doc.appendChild(rootElement)
            serializeNode(root, rootElement, doc)
            root.recycle()

            val transformer = TransformerFactory.newInstance().newTransformer()
            val result = StreamResult(StringWriter())
            transformer.transform(DOMSource(doc), result)
            result.writer.toString()
        } catch (e: Exception) {
            root.recycle()
            "<error>${e.message}</error>"
        }
    }

    private fun serializeNode(node: AccessibilityNodeInfo, parent: org.w3c.dom.Element, doc: org.w3c.dom.Document) {
        val element = doc.createElement("node")
        element.setAttribute("class", node.className?.toString() ?: "")
        element.setAttribute("text", node.text?.toString() ?: "")
        element.setAttribute("content-desc", node.contentDescription?.toString() ?: "")
        element.setAttribute("resource-id", node.viewIdResourceName ?: "")
        element.setAttribute("clickable", node.isClickable.toString())
        element.setAttribute("focusable", node.isFocusable.toString())
        element.setAttribute("scrollable", node.isScrollable.toString())
        element.setAttribute("checked", node.isChecked.toString())
        element.setAttribute("enabled", node.isEnabled.toString())
        element.setAttribute("password", node.isPassword.toString())
        element.setAttribute("selected", node.isSelected.toString())

        val bounds = Rect()
        node.getBoundsInScreen(bounds)
        element.setAttribute("bounds", "[${bounds.left},${bounds.top}][${bounds.right},${bounds.bottom}]")

        parent.appendChild(element)

        for (i in 0 until node.childCount) {
            val child = node.getChild(i) ?: continue
            serializeNode(child, element, doc)
            child.recycle()
        }
    }
}
